scene/gui/text_edit.cpp

```diff
@@ -1083,11 +1071,31 @@ void TextEdit::_notification(int p_what) {
 						char_margin = size.width - char_margin - TS->shaped_text_get_size(rid).x;
 					}
 
+					float char_w = font->get_char_size(' ', font_size).width;
 					for (int c = 0; c < carets.size(); c++) {
 						if (!clipped && has_selection(c) && line >= get_selection_from_line(c) && line <= get_selection_to_line(c)) { // Selection
 							int sel_from = (line > get_selection_from_line(c)) ? TS->shaped_text_get_range(rid).x : get_selection_from_column(c);
 							int sel_to = (line < get_selection_to_line(c)) ? TS->shaped_text_get_range(rid).y : get_selection_to_column(c);
 							Vector<Vector2> sel = TS->shaped_text_get_selection(rid, sel_from, sel_to);
+							// End of line selected
+							if (line < get_selection_to_line(c)) {
+								Vector<Vector2> char_bounds = TS->shaped_text_get_selection(rid, 0, str.length());
```

Comment:
Could we use something like `shaped_text_get_size` if we're calculating the entire length, rather then looping over everything?

Reply:
I didn't know this existed, thanks! It'll simplify the code nicely.